File: wisefy/src/androidTest/java/com/isupatches/wisefy/test/DisableWifiTests.java
package com.isupatches.wisefy.test;


import com.isupatches.wisefy.callbacks.DisableWifiCallbacks;
import com.isupatches.wisefy.constants.WiseFyCodes;
import com.isupatches.wisefy.test.base.BaseTestClass;
import com.isupatches.wisefy.test.base.TestActivity;
import org.junit.Test;
import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.anyBoolean;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.timeout;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;


public class DisableWifiTests extends BaseTestClass<TestActivity> {

    public DisableWifiTests() {
        super(TestActivity.class);
    }

    @Test
    public void noCallbacks_failure() {
        setManagers();
        when(mMockWiFiManager.setWifiEnabled(anyBoolean())).thenReturn(false);
        boolean result = mWiseFy.disableWifi();
        verify(mMockWiFiManager, timeout(2000)).setWifiEnabled(false);
        assertEquals(false, result);
    }

    @Test
    public void noCallbacks_failure_nullWifiManager() {
        setManagersToNull();
        boolean result = mWiseFy.disableWifi();
        assertEquals(false, result);
    }

    @Test
    public void noCallbacks_success() {
        setManagers();
        when(mMockWiFiManager.setWifiEnabled(anyBoolean())).thenReturn(true);
        boolean result = mWiseFy.disableWifi();
        verify(mMockWiFiManager, timeout(2000)).setWifiEnabled(false);
        assertEquals(true, result);
    }

    @Test
    public void callbacks_failure() {
        setManagers();
        when(mMockWiFiManager.setWifiEnabled(anyBoolean())).thenReturn(false);
        DisableWifiCallbacks mockDisableWifiCallbacks = mock(DisableWifiCallbacks.class);
        mWiseFy.disableWifi(mockDisableWifiCallbacks);
        verify(mMockWiFiManager, timeout(2000)).setWifiEnabled(false);
        verify(mockDisableWifiCallbacks, timeout(2000)).failureDisablingWifi();
    }

    @Test
    public void callbacks_failure_nullCallback() {
        setManagers();
        when(mMockWiFiManager.setWifiEnabled(anyBoolean())).thenReturn(false);
        DisableWifiCallbacks mockDisableWifiCallbacks = mock(DisableWifiCallbacks.class);
        mWiseFy.disableWifi(null);
        verify(mMockWiFiManager, timeout(2000)).setWifiEnabled(false);
        verify(mockDisableWifiCallbacks, never()).failureDisablingWifi();
    }

    @Test
    public void callbacks_failure_nullWifiManager() {
        setManagersToNull();
        DisableWifiCallbacks mockDisableWifiCallbacks = mock(DisableWifiCallbacks.class);
        mWiseFy.disableWifi(mockDisableWifiCallbacks);
        verify(mMockWiFiManager, never()).setWifiEnabled(false);
        verify(mockDisableWifiCallbacks, timeout(2000)).disableWifiWiseFyFailure(WiseFyCodes.MISSING_PREREQUISITE);
    }

    @Test
    public void callbacks_failure_nullWifiManager_nullCallback() {
        setManagersToNull();
        DisableWifiCallbacks mockDisableWifiCallbacks = mock(DisableWifiCallbacks.class);
        mWiseFy.disableWifi(null);
        verify(mMockWiFiManager, never()).setWifiEnabled(false);
        verify(mockDisableWifiCallbacks, never()).disableWifiWiseFyFailure(WiseFyCodes.MISSING_PREREQUISITE);
    }

    @Test
    public void callbacks_success() {
        setManagers();
        when(mMockWiFiManager.setWifiEnabled(anyBoolean())).thenReturn(true);
        DisableWifiCallbacks mockDisableWifiCallbacks = mock(DisableWifiCallbacks.class);
        mWiseFy.disableWifi(mockDisableWifiCallbacks);
        verify(mMockWiFiManager, timeout(2000)).setWifiEnabled(false);
        verify(mockDisableWifiCallbacks, timeout(2000)).wifiDisabled();
    }

    @Test
    public void callbacks_success_nullCallback() {
        setManagers();
        when(mMockWiFiManager.setWifiEnabled(anyBoolean())).thenReturn(true);
        DisableWifiCallbacks mockDisableWifiCallbacks = mock(DisableWifiCallbacks.class);
        mWiseFy.disableWifi(null);
        verify(mMockWiFiManager, timeout(2000)).setWifiEnabled(false);
        verify(mockDisableWifiCallbacks, never()).wifiDisabled();
    }
}

File: wisefy/src/test/java/com/isupatches/wisefy/DisableWifiTests.kt
package com.isupatches.wisefy

import android.os.Build

import com.isupatches.wisefy.callbacks.DisableWifiCallbacks
import com.isupatches.wisefy.constants.MISSING_PARAMETER
import com.isupatches.wisefy.internal.base.BaseUnitTest

import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith

import org.mockito.Mockito.mock
import org.mockito.Mockito.timeout
import org.mockito.Mockito.verify

import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/**
 * Tests the ability to disable a device's wifi.
 *
 * @author Patches
 */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [Build.VERSION_CODES.M])
internal class DisableWifiTests : BaseUnitTest() {

    @Test fun sync_failure_prechecks() {
        mockWiseFyPrechecksUtil.disableWifi_failure()
        assertEquals(false, wisefy.disableWifi())
        verificationUtil.didNotTryToDisableWifi()
    }

    @Test fun sync_failure() {
        mockWiseFyPrechecksUtil.disableWifi_success()
        mockNetworkUtil.disableWifi_failure()
        assertEquals(false, wisefy.disableWifi())
        verificationUtil.triedToDisableWifi()
    }

    @Test fun sync_success() {
        mockWiseFyPrechecksUtil.disableWifi_success()
        mockNetworkUtil.disableWifi_success()
        assertEquals(true, wisefy.disableWifi())
        verificationUtil.triedToDisableWifi()
    }

    @Test fun async_failure_prechecks() {
        mockWiseFyPrechecksUtil.disableWifi_failure()
        val mockCallbacks = mock(DisableWifiCallbacks::class.java)
        wisefy.disableWifi(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wisefyFailure(MISSING_PARAMETER)
        verificationUtil.didNotTryToDisableWifi()
    }

    @Test fun async_failure_prechecks_nullCallback() {
        mockWiseFyPrechecksUtil.disableWifi_failure()
        nullCallbackUtil.callDisableWifi()
        verificationUtil.didNotTryToDisableWifi()
    }

    @Test fun async_failure() {
        mockWiseFyPrechecksUtil.disableWifi_success()
        mockNetworkUtil.disableWifi_failure()
        val mockCallbacks = mock(DisableWifiCallbacks::class.java)
        wisefy.disableWifi(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).failureDisablingWifi()
        verificationUtil.triedToDisableWifi()
    }

    @Test fun async_failure_nullCallback() {
        mockWiseFyPrechecksUtil.disableWifi_success()
        mockNetworkUtil.disableWifi_failure()
        nullCallbackUtil.callDisableWifi()
        verificationUtil.triedToDisableWifi()
    }

    @Test fun async_success() {
        mockWiseFyPrechecksUtil.disableWifi_success()
        mockNetworkUtil.disableWifi_success()
        val mockCallbacks = mock(DisableWifiCallbacks::class.java)
        wisefy.disableWifi(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wifiDisabled()
        verificationUtil.triedToDisableWifi()
    }

    @Test fun async_success_nullCallback() {
        mockWiseFyPrechecksUtil.disableWifi_success()
        mockNetworkUtil.disableWifi_success()
        nullCallbackUtil.callDisableWifi()
        verificationUtil.triedToDisableWifi()
    }
}
